_posts/2018-03-16-first-steps-into-gen-servers.md

```diff
@@ -0,0 +1,77 @@
+
+A big part of my growth as an engineer comes from getting out of my comfort zone, using new and unfamiliar tools to build hopefully useful tools. After doing a little research on potential solutions for our automation opportunity, I settled on Elixir's GenServer, or Generic Server, pattern. Coming from a primarily Object Oriented Programming background with Rails, this pattern was a bit foreign to me, to say the least.
+
+### Getting Started
+
+The basic premise for this project was to spin up a simple Phoenix/Elixir microservice application to perform two main tasks -- make an API call to one of our monolith apps for portfolio data and then generate an email with the results once a month. After building out the API call in Rails, it was time to get started on the GenServer.
+
+From the [GenServer](https://hexdocs.pm/elixir/GenServer.html) HexDocs:
+
+> A GenServer is a process like any other Elixir process and it can be used to keep state, execute code asynchronously and so on.
+
+Seems pretty straightforward. By writing a GenServer and isolating this single task into its own process, I am effectively creating a microservice that can be managed by the application's own [supervision tree](https://hexdocs.pm/elixir/Supervisor.html).
+
+Here's what I ended up with, using the default callbacks for GenServer:
+
+```
+  # Default implementation of GenServer
+  def start_link(opts) do
+    GenServer.start_link(__MODULE__, :ok, opts)
+  end
+
+  # Initializes the GenServer on application startup
+  def init(state) do
+    schedule_work()
+    {:ok, state}
+  end
+
+  # Generic catch-all message/event handler
+  def handle_info(:work, state) do
+    do_work() # The actual business logic, defined below
+    schedule_work() # Process scheduler
+    {:noreply, state}
+  end
+
+  # The work!
+  def do_work() do
+    # Business logic goes here!
+  end
+
+  # Scheduler to set interval for do_work()
+  defp schedule_work() do
+    Process.send_after(self(), :work, INTERVAL_IN_MILLESECONDS)
+  end
+```
+
+Back to the docs to explain what's going on here:
+> The goal of a GenServer is to abstract the “receive” loop for developers, automatically handling system messages, support code change, synchronous calls and more.
+
+Once this GenServer gets integrated into the supervision tree,  we've essentially created a black box that will receive periodic messages to "do work".
+
+No controllers. No models. No database.
+
+Doesn't get much simpler than that!
+
+### Future Implications
+
+At this point, you know that I can now read and follow basic documentation. What a relief!
+
+What I find most fascinating about this particular pattern is its ability to effectively manage multiple microservices within a single supervision tree. While in this particular example I'm only implementing a single GenServer, adding another two, three...10, more, would be a trivial process. And they could even communicate with each other with their own messages.
```

Comment:
I think you can drop the commas before and after 'more'? I didn't go to j-school so correct me if I'm wrong ;)

also, 'communicate with each other with their own messages' is a little awkward with the double 'with'. Maybe 'using their own messages'?

Reply:
👍